hydradx/tests/test_omnipool_amm.py

```diff
@@ -1,306 +1,301 @@
 
-def test_prices(d):
-    for i in range(len(d['Q'])):
-        assert oamm.price_i(d, i) > 0
+    i = old_state.asset_list[-1]
+    old_state.weight_cap[i] = old_state.tvl[i] / old_state.tvl_total * 1.00001
 
+    # calculate what should be the maximum allowable liquidity provision
+    max_amount = ((old_state.weight_cap[i] / (1 - old_state.weight_cap[i])
+                  * old_state.lrna_total - old_state.lrna[i] / (1 - old_state.weight_cap[i]))
+                  / oamm.price_i(old_state, i))
 
-def test_initialize_pool_state(d):
-    pass  # TODO
+    # make sure agent has enough funds
+    old_agents[lp_id]['r'][i] = max_amount * 2
 
+    # try one just above and just below the maximum allowable amount
+    illegal_state, illegal_agents = oamm.add_risk_liquidity(old_state, old_agents, lp_id, max_amount * 1.0001, i)
+    if illegal_state.liquidity[i] != pytest.approx(old_state.liquidity[i]):
+        raise AssertionError(f'illegal transaction passed against weight limit in {i}')
+    legal_state, legal_agents = oamm.add_risk_liquidity(old_state, old_agents, lp_id, max_amount * 0.9999, i)
+    if legal_state.liquidity[i] == old_state.liquidity[i]:
+        raise AssertionError(f'legal transaction failed against weight limit in {i}')
 
-@given(QR_strat)
-def test_QR_strat(d):
-    test_prices(d)
-    test_initialize_pool_state(d)
 
+@given(omnipool_config())
+def test_QR_strat(market_state: oamm.OmnipoolState):
+    for i in market_state.asset_list:
+        assert oamm.price_i(market_state, i) > 0
 
-@given(QR_strat)
-def test_add_risk_liquidity(initial_state):
-    token_count = len(initial_state['R'])
-    old_state = oamm.state_dict(
-        token_list=['HDX', 'USD'] + ['token'] * (token_count-2),
-        r_values=initial_state['R'],
-        s_values=[1500000] * token_count,
-        p_values=[oamm.price_i(initial_state, j) for j in range(token_count)],
-        omega_values=[0.5] * token_count
-    )
 
+@given(omnipool_config(token_count=3, lrna_fee=0, asset_fee=0))
+def test_add_risk_liquidity(initial_state: oamm.OmnipoolState):
+    old_state = initial_state
     LP_id = 'LP'
+    i = old_state.asset_list[2]
+
     old_agents = {
         LP_id: {
-            'r': [1000, 0],
-            's': [0, 0],
-            'p': [None, None]
+            'r': {i: 1000},
+            's': {i: 0},
+            'p': {i: old_state.liquidity[i] / old_state.lrna[i]}
         }
     }
     delta_R = 1000
-    i = 0
 
     new_state, new_agents = oamm.add_risk_liquidity(old_state, old_agents, LP_id, delta_R, i)
-    for j in range(len(['R'])):
+    for j in initial_state.asset_list:
         assert oamm.price_i(old_state, j) == pytest.approx(oamm.price_i(new_state, j))
-    assert old_state['R'][i] / old_state['S'][i] == pytest.approx(new_state['R'][i] / new_state['S'][i])
+    if old_state.liquidity[i] / old_state.shares[i] != pytest.approx(new_state.liquidity[i] / new_state.shares[i]):
+        raise AssertionError(f'Price change in {i}'
```

Comment:
Yes, the reason I keep changing that is because with a plain assert, when I run the tests one at a time, it fails without giving me any information about why. When I do it this way, I can put a breakpoint on the raise statement and see exactly why it failed, and even rewind and watch it fail. Easier to debug. I also looked into it and found that the only difference between 'assert condition x' and 'if condition x: raise assertionError' is that the former will be ignored during compile if you turn a certain optimization on. Otherwise they are the equivalent (except the one is easier to debug).

Reply:
ah so it plays well with hypothesis? If so, sounds like an improvement.

Reply:
Yes, hypothesis has no problem with this.